VisualStudio 2013 can't compile SPIRV-Tools

There is still interest in building SPIRV-Tools with VisualStudio 2013.  VS 2013 has known bugs in its C++11 support, so we would have to work around those to make it build.
We would have to exclude optional components used in testing -- effcee and RE2 -- since effcee depends on RE2 and RE2 rejects VS 2013.

I'm attaching a build log with VS2013.
[vs2013.txt](https://github.com/KhronosGroup/SPIRV-Tools/files/1690925/vs2013.txt)

Most of the errors are the same.  Filtering for "error C" I get the following:

C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_92ee5ab7ae9c27d29b5884b4d7be43e3>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_563a50b858269031fca6726f74fd14d7>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_e8cce036caa56f0380b9a8239f1cdc59>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_4d2ba9ba9d3194532607343704422331>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_d2f5e68a0e1f8e93a9925548676f0acc>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_eac27fe7d4ae8ba77409d99d3d5dae15>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_37b3e924f6024c89cb5b8d4f9784b035>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_42529a2492979c5a4991101c21a5d1fe>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_cdd70d38fd2d9944c0f155cdc68c5984>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_643255804eaf28bf95a4e19afe1acfde>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_93b055be0a0aea8176f981f93cd6a649>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::Loop::FindLoopPreheader::<lambda_ee66ba5b98bc7f60d53acaa2e3f3cf07>::operator ()(const uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'const uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_4cafed0d858d0c7bf6a9fc7b8aaab8fd>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_f20d9ac4004c06ab3e4109f1c671cd60>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::opt::InlinePass::UpdateSucceedingPhis::<lambda_cd2042a6f941db7182bab4b5e7449d97>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::opt::InlinePass::ComputeStructuredSuccessors::<lambda_d4cf9d5354c413708287f3b33bd7b72d>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xrefwrap(283) : error C2664: 'void spvtools::ir::CFG::AddEdges::<lambda_eae2f40f867a0f9a7f457ce07e6ab4f8>::operator ()(uint32_t) const' : cannot convert argument 1 from 'unsigned int *' to 'uint32_t'

When mangling the code at propagator.cpp to look like this:

    const std::function<void(const uint32_t)> doit = [](const uint32_t) {};
    &block.ForEachSuccessorLabel(doit);

I get this error:
C:\spirv-tools\SPIRV-Tools\source\opt\propagator.cpp(213) : error C2666: 'spvtools::ir::BasicBlock::ForEachSuccessorLabel' : 2 overloads have similar conversions
        c:\spirv-tools\spirv-tools\source\opt\basic_block.h(139): could be 'void spvtools::ir::BasicBlock::ForEachSuccessorLabel(const std::function<void (uint32_t *)> &)'
        c:\spirv-tools\spirv-tools\source\opt\basic_block.h(133): or       'void spvtools::ir::BasicBlock::ForEachSuccessorLabel(const std::function<void (uint32_t)> &) const'
        while trying to match the argument list '(const std::function<void (uint32_t)>)'
        note: qualification adjustment (const/volatile) may be causing the ambiguity

So VS2013 is not seeing inside the argument list when overloading.

--- a/source/opt/loop_descriptor.cpp
+++ b/source/opt/loop_descriptor.cpp
@@ -296,7 +296,11 @@ bool Loop::IsLCSSA() const {
   return true;
 }
 
-LoopDescriptor::LoopDescriptor(const Function* f) { PopulateList(f); }
+LoopDescriptor::LoopDescriptor(const Function* f) : loops_() {
+  PopulateList(f);
+}
+
+LoopDescriptor::~LoopDescriptor() { ClearLoops(); }
 
 void LoopDescriptor::PopulateList(const Function* f) {
   IRContext* context = f->GetParent()->context();
@@ -304,7 +308,7 @@ void LoopDescriptor::PopulateList(const Function* f) {
   opt::DominatorAnalysis* dom_analysis =
       context->GetDominatorAnalysis(f, *context->cfg());
 
-  loops_.clear();
+  ClearLoops();
 
   // Post-order traversal of the dominator tree to find all the OpLoopMerge
   // instructions.
@@ -329,14 +333,14 @@ void LoopDescriptor::PopulateList(const Function* f) {
       BasicBlock* header_bb = context->get_instr_block(merge_inst);
 
       // Add the loop to the list of all the loops in the function.
-      loops_.emplace_back(MakeUnique<Loop>(context, dom_analysis, header_bb,
-                                           continue_bb, merge_bb));
-      Loop* current_loop = loops_.back().get();
+      Loop* current_loop =
+          new Loop(context, dom_analysis, header_bb, continue_bb, merge_bb);
+      loops_.push_back(current_loop);
 
       // We have a bottom-up construction, so if this loop has nested-loops,
       // they are by construction at the tail of the loop list.
       for (auto itr = loops_.rbegin() + 1; itr != loops_.rend(); ++itr) {
-        Loop* previous_loop = itr->get();
+        Loop* previous_loop = *itr;
 
         // If the loop already has a parent, then it has been processed.
         if (previous_loop->HasParent()) continue;
@@ -364,9 +368,16 @@ void LoopDescriptor::PopulateList(const Function* f) {
       }
     }
   }
-  for (std::unique_ptr<Loop>& loop : loops_) {
-    if (!loop->HasParent()) dummy_top_loop_.nested_loops_.push_back(loop.get());
+  for (Loop* loop : loops_) {
+    if (!loop->HasParent()) dummy_top_loop_.nested_loops_.push_back(loop);
   }
+}
+
+void LoopDescriptor::ClearLoops() {
+  for (Loop* loop : loops_) {
+    delete loop;
+  }
+  loops_.clear();
 }
 
 }  // namespace ir

--- a/source/opt/loop_descriptor.h
+++ b/source/opt/loop_descriptor.h
@@ -256,6 +256,21 @@ class LoopDescriptor {
   // Creates a loop object for all loops found in |f|.
   explicit LoopDescriptor(const Function* f);
 
+  // Disable copy constructor, to avoid double-free on destruction.
+  LoopDescriptor(const LoopDescriptor&) = delete;
+  // Move constructor.
+  LoopDescriptor(LoopDescriptor&& other) {
+    // We need to take ownership of the Loop objects in the other
+    // LoopDescriptor, to avoid double-free.
+    loops_ = std::move(other.loops_);
+    other.loops_.clear();
+    basic_block_to_loop_ = std::move(other.basic_block_to_loop_);
+    other.basic_block_to_loop_.clear();
+  }
+
+  // Destructor
+  ~LoopDescriptor();
+
   // Returns the number of loops found in the function.
   inline size_t NumLoops() const { return loops_.size(); }
 
@@ -264,7 +279,7 @@ class LoopDescriptor {
   inline Loop& GetLoopByIndex(size_t index) const {
     assert(loops_.size() > index &&
            "Index out of range (larger than loop count)");
-    return *loops_[index].get();
+    return *loops_[index];
   }
 
   // Returns the inner most loop that contains the basic block id |block_id|.
@@ -296,7 +311,9 @@ class LoopDescriptor {
   }
 
  private:
-  using LoopContainerType = std::vector<std::unique_ptr<Loop>>;
+  // TODO(dneto): This should be a vector of unique_ptr.  But VisualStudio 2013
+  // is unable to compile it.
+  using LoopContainerType = std::vector<Loop*>;
 
   // Creates loop descriptors for the function |f|.
   void PopulateList(const Function* f);
@@ -308,7 +325,11 @@ class LoopDescriptor {
     return it != basic_block_to_loop_.end() ? it->second : nullptr;
   }
 
-  // A list of all the loops in the function.
+  // Erase all the loop information.
+  void ClearLoops();
+
+  // A list of all the loops in the function.  This variable owns the Loop
+  // objects.
   LoopContainerType loops_;
   // Dummy root: this "loop" is only there to help iterators creation.
   Loop dummy_top_loop_;

